pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java

```diff
@@ -0,0 +1,590 @@
+                    // wire is prefixed with the number of bytes for the message. However, to prevent a DOS attack
+                    // where the attacker sends us a very large length and exhausts our memory, we have a maximum
+                    // message size configuration setting. Using that, we can detect attempts to exhaust our memory.
+                    final long length = data.readUnsignedInt32();
+                    if (length > config.maxMessageSizeBytes()) {
+                        throw new GrpcException(GrpcStatus.INVALID_ARGUMENT,
+                                "Message size exceeds maximum allowed size");
+                    }
+                    // Create a buffer to hold the message. We sadly cannot reuse this buffer because once we have
+                    // filled it and wrapped it in Bytes and sent it to the handler, some user code may grab and hold
+                    // that Bytes object for an arbitrary amount of time, and if we were to scribble into the same
+                    // byte array, we would break the application. So we need a new buffer each time :-(
+                    entityBytes = new byte[(int) length];
+                    entityBytesIndex = 0;
+                }
+
+                // By the time we get here, entityBytes is no longer null. It may be empty, or it may already have
+                // been partially populated from a previous iteration. It may be that the number of bytes available
+                // to be read is larger than just this one message. So we need to be careful to read, from what is
+                // available, only up to the message length, and to leave the rest for the next iteration.
+                final int available = data.available();
+                final int numBytesToRead = Math.min(entityBytes.length - entityBytesIndex, available);
+                data.read(entityBytes, entityBytesIndex, numBytesToRead);
+                entityBytesIndex += numBytesToRead;
+
+                // If we have completed reading the message, then we can proceed.
+                if (entityBytesIndex == entityBytes.length) {
+                    // Grab and wrap the bytes and reset to being reading the next message
+                    final var bytes = Bytes.wrap(entityBytes);
+                    incoming.onNext(bytes);
+                    entityBytesIndex = 0;
+                    entityBytes = null;
+                }
+            }
+
+            // The end of the stream has been reached! It is possible that a bad client will send end of stream before
+            // all the message data we sent. In that case, it is as if the message were never sent.
+            if (header.flags(Http2FrameTypes.DATA).endOfStream()) {
+                entityBytesIndex = 0;
+                entityBytes = null;
+                currentStreamState = Http2StreamState.HALF_CLOSED_LOCAL;
+                incoming.onComplete();
+            }
```

Comment:
What about the `else` branch? Should it re-initialize the variables? Should it log an error of some sort? Or emit a metric? Or maybe keep spinning the loop above? Or tell the `incoming` anything?  Or...?


BTW, should the loop above exit gracefully if the state changes while it's spinning (like if someone calls close() from another thread or something?)

Reply:
> What about the else branch

I don't think any else branch is necessary. The `else` is the "standard" case, where the method call isn't complete, and we're just waiting on more data from the caller. In this case, the `data` method is going to be called again at some point in the future, and we need to continue holding these variables to keep track of what we've read so far.

Reply:
Ah, I missed the idea of the `data()` method being called multiple times for the same request. This makes sense then. Thanks for the clarification.